PAAS-1322: Remove envLink on Jahia env deletion

amitc-jahia opened the pull request:
JIRA issue: https://jira.jahia.org/browse/PAAS-1322

Short description:

kevinvandel commented on install.yml:
```diff
@@ -336,8 +336,32 @@ onAfterClone:
+                                return object.name == "cp";
+                              }).pop().envLink;
+        if (jcustomer_env != null) {
+          envs = jelastic.env.control.GetNodeGroups(envName, session).object.filter(function (object) {
```
Shouldn't this be run on "jcustomer_env" instead of "envName" ?

kevinvandel commented on install.yml:
```diff
@@ -336,8 +336,32 @@ onAfterClone:
+  deleteEnvLink:
+    - script: |-
+        const envName = "${env.envName}";
+        jcustomer_env = jelastic.env.control.GetNodeGroups(envName, session).object.filter(function (object) {
```
It would be nice to keep it to camelCase (or at least not mixing different cases, but since this is JavaScript camelCase would be better)

amitc-jahia commented:
Changes made. Please review again.

kevinvandel commented on install.yml:
```diff
@@ -336,8 +336,37 @@ onAfterClone:
+        jCustomerEnv = jelastic.env.control.GetNodeGroups(envName, session).object.filter(function (object) {
+                                return object.name == "cp";
+                              }).pop().envLink;
+        if (jCustomerEnv != null) {
```
To avoid having a useless `else` block, here we can do this:
```
        if (jCustomerEnv == null) {
          return {"result": 0, "out": "No jCustomer env linked"};
        }
```
And continue without any `else` block (useless because of the `return` statement).

Generally speaking, this is a best practice to check particular cases like that. This avoids having a lot of `if` blocks piling up until we get to the "standard" case, and a cascade of `else` blocks at the end handling particular cases.

kevinvandel commented on install.yml:
```diff
@@ -336,8 +336,37 @@ onAfterClone:
+                              }).pop().envLink;
+          if (envsLinked.indexOf(",") != -1) {
+            envsLinkedArr = envsLinked.split(",");
+            removedEnv = envsLinkedArr.splice(envsLinkedArr.indexOf[envName], 1);
```
This doesn't work at all because you used square brackets instead of parentheses. Did you really test it ? It took me 30 seconds with JSFiddle to see that nothing was working as expected.
```suggestion
            removedEnv = envsLinkedArr.splice(envsLinkedArr.indexOf(envName), 1);
```

kevinvandel replied:
Also, we need to check `envsLinkedArr` really contains `envName` (i.e. index isn't `-1`), even though it shouldn't happen. But if for whatever reason we have an issue and the Jahia environment was not present in the jCustomer environment `envLink`, this script will randomly remove the last item of the `envLink` list, causing even more trouble (that we will probably have a hard time figuring out).

kevinvandel commented on install.yml:
```diff
@@ -336,8 +336,37 @@ onAfterClone:
+          if (envsLinked.indexOf(",") != -1) {
+            envsLinkedArr = envsLinked.split(",");
+            removedEnv = envsLinkedArr.splice(envsLinkedArr.indexOf[envName], 1);
+            newEnvLink = String(envsLinkedArr);
+            resp = jelastic.env.control.ApplyNodeGroupData(jCustomerEnv, session, nodeGroup='cp', data={'envLink': newEnvLink});
+            return {"result": 0, "out": "Jahia env " + removedEnv + " removed from envLink of " + jCustomerEnv};
+          } else {
+              resp = jelastic.env.control.ApplyNodeGroupData(jCustomerEnv, session, nodeGroup='cp', data={'envLink': null});
+              return {"result": 0, "out": "Jahia env " + envName + " removed from envLink of " + jCustomerEnv};
+          }
```
We can avoid duplicating code like this (also fixes indenting issue):
```suggestion
            newEnvLink = String(envsLinkedArr);
          } else {
            newEnvLink = null;
          }
          resp = jelastic.env.control.ApplyNodeGroupData(jCustomerEnv, session, nodeGroup='cp', data={'envLink': newEnvLink});
          return {"result": 0, "out": "Jahia env " + envName + " removed from envLink of " + jCustomerEnv};
```

kevinvandel commented on install.yml:
```diff
@@ -336,8 +336,37 @@ onAfterClone:
+          envsLinked = jelastic.env.control.GetNodeGroups(jCustomerEnv, session).object.filter(function (object) {
+                                return object.name == "cp";
+                              }).pop().envLink;
+          if (envsLinked.indexOf(",") != -1) {
```
It could be nice to add a little comment here, because the `indexOf(",")` is not really explicit, especially if you don't have in mind that on jCustomer side the `envLink` can be a comma-separated list.

amitc-jahia commented:
Following changes made as requested : 

- Replaced the square brackets with parenthesis
- Added `if` condition to check whether jCustomer is `null` and continue without `else` block if condition is not met
- Check for the `envName` if present in the `envsLinked` of jCustomer
- Removed duplication of code
- Added comment to highlight that envLink on jCustomer side can contain multiple Jahia envs

Please review.

kevinvandel commented on install.yml:
```diff
@@ -336,8 +336,38 @@ onAfterClone:
+        }
+        if (envsLinked.indexOf(",") != -1) {      // envLink can contain multiple Jahia envs on jCustomer side
+          envsLinkedArr = envsLinked.split(",");
+          removedEnv = envsLinkedArr.splice(envsLinkedArr.indexOf(envName), 1);
```
`removedEnv` is never used, it can be removed:
```suggestion
          envsLinkedArr.splice(envsLinkedArr.indexOf(envName), 1);
```

kevinvandel commented on install.yml:
```diff
@@ -336,8 +336,38 @@ onAfterClone:
+          removedEnv = envsLinkedArr.splice(envsLinkedArr.indexOf(envName), 1);
+          newEnvLink = String(envsLinkedArr);
+        } else {
+            newEnvLink = null;
```
Indenting issue (4 spaces here, while you use 2 spaces everywhere else)

kevinvandel commented on install.yml:
```diff
@@ -336,8 +336,38 @@ onAfterClone:
+        if (envsLinked.indexOf(envName) == -1) {
+          return {"result": 0, "out": envName + " not in envLink of " + jCustomerEnv};
+        }
+        if (envsLinked.indexOf(",") != -1) {      // envLink can contain multiple Jahia envs on jCustomer side
```
To avoid having an inline comment on the same line as the `if`, could you please put it on a new line before the `if` ?
Also in the comment it would be nice to specify that in this case values are separated with commas (so that we clearly understand the `indexOf(",")` thing).

amitc-jahia commented:
Changes made as requested. Please review.
